distributed/tests/test_priorities.py

```diff
@@ -1,128 +1,183 @@
+        await ev.set()
+        await wait(clog)
+        await wait(low)
 
 
-@gen_cluster(client=True)
-async def test_annotate_persist(c, s, a, b):
+@gen_cluster(client=True, nthreads=[("", 1)])
+async def test_annotate_persist(c, s, a):
     with dask.annotate(priority=-1):
-        low = delayed(inc)(1, dask_key_name="low")
+        low = dinc(1, dask_key_name="low")
+    ev = Event()
+    clog = dwait(ev, dask_key_name="clog")
     with dask.annotate(priority=1):
-        high = delayed(inc)(2, dask_key_name="high")
-    many = [delayed(slowinc)(i, delay=0.1) for i in range(4)]
-
-    low, high, x, y, z, w = persist(low, high, *many, optimize_graph=False)
-    await wait(high)
-    assert s.tasks[low.key].state == "processing"
-
-
-@gen_cluster(client=True, nthreads=[("127.0.0.1", 1)])
-async def test_repeated_persists_same_priority(c, s, w):
-    xs = [delayed(slowinc)(i, delay=0.05, dask_key_name="x-%d" % i) for i in range(10)]
-    ys = [
-        delayed(slowinc)(x, delay=0.05, dask_key_name="y-%d" % i)
-        for i, x in enumerate(xs)
-    ]
-    zs = [
-        delayed(slowdec)(x, delay=0.05, dask_key_name="z-%d" % i)
-        for i, x in enumerate(xs)
-    ]
-
-    ys = dask.persist(*ys)
-    zs = dask.persist(*zs)
-
-    while (
-        sum(t.state == "memory" for t in s.tasks.values()) < 5
-    ):  # TODO: reduce this number
+        high = dinc(2, dask_key_name="high")
+    low, clog, high = c.persist([low, clog, high], optimize_graph=False)
+
+    while len(s.tasks) < 3:
+        await asyncio.sleep(0.01)
+
+    async with Worker(s.address, nthreads=1):  # See comment in test_submit
+        await wait(high)
+        assert s.tasks[low.key].state == "processing"
+        await ev.set()
+        await wait(clog)
+        await wait(low)
+
+
+@gen_cluster(client=True, nthreads=[])
+async def test_repeated_persists_same_priority(c, s):
+    xs = [dinc(i, dask_key_name=f"x{i}") for i in range(3)]
+    ys = [dinc(xs[i], dask_key_name=f"y{i}") for i in range(3)]
+    zs = [dinc(i, dask_key_name=f"z{i}") for i in range(3)]
+    ys = c.persist(ys)
+    zs = c.persist(zs)
+
+    while len(s.tasks) < 9:
         await asyncio.sleep(0.01)
 
-    assert any(s.tasks[y.key].state == "memory" for y in ys)
-    assert any(s.tasks[z.key].state == "memory" for z in zs)
+    async with Worker(s.address, nthreads=1) as a:  # See comment in test_submit
+        await wait(ys + zs)
+        story = a.story("x0", "x1", "x2", "y0", "y1", "y2", "z0", "z1", "z2")
+
+    order = [ev[0] for ev in story if ev[1:3] == ("ready", "executing")]
+    # e.g. order = ['x0', 'z0', 'z1', 'z2', 'x1', 'x2', 'y0', 'y1', 'y2']
+    # There's an element of randomness, but there must be *some* interleaving of x and z
+    assert order[-3:] == ["y0", "y1", "y2"]
+    assert order.index("x0") < order.index("x1") < order.index("x2")
+    assert order.index("z0") < order.index("z1") < order.index("z2")
+    assert order.index("x0") < order.index("z2")
```

Comment:
I honestly considered the previous version more easy to read and maintain. changing the 5 to a 9 in the `sum(t.state ...)` condition made the test pass for me 1000 times while the 5 indeed was flaky.

This version even fails for me locally.